lib/schema/uuid.js

```diff
@@ -0,0 +1,300 @@
+ * @inherits SchemaType
+ * @api public
+ */
+
+function SchemaUUID(key, options) {
+  SchemaType.call(this, key, options, 'UUID');
+  this.getters.push(binaryToString);
+}
+
+/**
+ * This schema type's name, to defend against minifiers that mangle
+ * function names.
+ *
+ * @api public
+ */
+SchemaUUID.schemaName = 'UUID';
+
+SchemaUUID.defaultOptions = {};
+
+/*!
+ * Inherits from SchemaType.
+ */
+SchemaUUID.prototype = Object.create(SchemaType.prototype);
+SchemaUUID.prototype.constructor = SchemaUUID;
+
+/*!
+ * ignore
+ */
+
+SchemaUUID._cast = function(value) {
+  if (value === null) {
+    return value;
+  }
+
+  function newBuffer(initbuff) {
+    const buff = new MongooseBuffer(initbuff);
+    buff._subtype = 4;
+    return buff;
+  }
+
+  if (typeof value === 'string') {
+    if (UUID_FORMAT.test(value)) {
+      return stringToBinary(value);
+    } else {
+      throw new CastError(SchemaUUID.schemaName, value, this.path);
+    }
+  }
+
+  if (Buffer.isBuffer(value)) {
+    return newBuffer(value);
+  }
+
+  if (value instanceof Binary) {
+    return newBuffer(value.value(true));
+  }
+
+  // Re: gh-647 and gh-3030, we're ok with casting using `toString()`
+  // **unless** its the default Object.toString, because "[object Object]"
+  // doesn't really qualify as useful data
+  if (value.toString && value.toString !== Object.prototype.toString) {
+    if (UUID_FORMAT.test(value.toString())) {
+      return stringToBinary(value.toString());
+    }
+  }
+
+  throw new CastError(SchemaUUID.schemaName, value, this.path);
+};
+
+/**
+ * Sets a default option for all UUID instances.
+ *
+ * #### Example:
+ *
+ *     // Make all UUIDs have `required` of true by default.
+ *     mongoose.Schema.UUID.set('required', true);
+ *
+ *     const User = mongoose.model('User', new Schema({ test: mongoose.UUID }));
+ *     new User({ }).validateSync().errors.test.message; // Path `test` is required.
+ *
+ * @param {String} option The option you'd like to set the value for
+ * @param {Any} value value for option
+ * @return {undefined}
+ * @function set
+ * @static
+ * @api public
+ */
+
+SchemaUUID.set = SchemaType.set;
+
+/**
+ * Get/set the function used to cast arbitrary values to UUIDs.
+ *
+ * #### Example:
+ *
+ *     // Make Mongoose refuse to cast UUIDs with 0 length
+ *     const original = mongoose.Schema.Types.UUID.cast();
+ *     mongoose.UUID.cast(v => {
+ *       assert.ok(typeof v === "string" && v.length > 0);
+ *       return original(v);
+ *     });
+ *
+ *     // Or disable casting entirely
+ *     mongoose.UUID.cast(false);
+ *
+ * @param {Function} [caster]
+ * @return {Function}
+ * @function get
+ * @static
+ * @api public
+ */
+
+SchemaUUID.cast = function cast(caster) {
+  if (arguments.length === 0) {
+    return this._cast;
+  }
+  if (caster === false) {
+    caster = this._defaultCaster;
+  }
+  this._cast = caster;
+
+  return this._cast;
+};
+
+/*!
+ * ignore
+ */
+
+SchemaUUID._checkRequired = v => UUID_FORMAT.test(v);
```

Comment:
Not necessary to do `UUID_FORMAT.test(v)`, just do `v != null`. This implementation will lead to Mongoose checking against UUID_FORMAT twice.

Reply:
i had originally used Decimal128 as a template, but i did not find why `_checkRequired` as set when `checkRequired` was used, if it is the same as `_cast` and `cast`, then it should have not much effect on the current implementation because `checkRequired` does not call `_checkRequired`

Reply:
resolved in c76d09784cc95b61f79554218712fa25e818cb43